llvm/include/llvm/Object/SYCLBIN.h

```diff
@@ -0,0 +1,86 @@
+//===- SYCLBIN.h - SYCLBIN binary format support ----------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_OBJECT_SYCLBIN_H
+#define LLVM_OBJECT_SYCLBIN_H
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Object/Binary.h"
+#include "llvm/SYCLLowerIR/ModuleSplitter.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include <string>
+
+namespace llvm {
+
+namespace object {
+
+class SYCLBIN : public Binary {
+public:
+  SYCLBIN(MemoryBufferRef Source);
+
+  enum class BundleState : uint8_t { Input = 0, Object = 1, Executable = 2 };
+  enum class IRType : uint8_t { SPIRV = 0, PTX = 1, AMDGCN = 2 };
+
+  struct ModuleDesc {
+    BundleState State;
+    std::string ArchString;
+    std::vector<module_split::SplitModule> SplitModules;
+  };
+
+  /// The current version of the binary used for backwards compatibility.
+  static constexpr uint32_t Version = 1;
+
+  /// Magic number used to identify SYCLBIN files.
+  static constexpr uint8_t MagicNumber[4] = {0x53, 0x59, 0x42, 0x49};
+
+  /// Serialize the contents of \p ModuleDescs to a binary buffer to be read
+  /// later.
+  static Expected<SmallString<0>> write(const SmallVector<ModuleDesc> &);
+
+  static Expected<std::unique_ptr<SYCLBIN>> read(MemoryBufferRef Source);
```

Comment:
Implementation of read/write is quite complex, it has to be exhaustively tested. I'm not saying that all the testing should be here from day 1, but we at least need to lay some groundwork for it. For example what you can do is to have a unit-test which does `write`, then `read` and checks that result is the same as the initial input. See e471ba3d0122d4c6601029d81c385cb6ebb9d7a4 for inspiration.

Reply:
Unittest has been added. Definitely could use some expanding on, but there are also a handful of TODOs in the current version, so it will be expanded upon over time.